File: src/lib/audio_track.dart
import 'dart:typed_data';

class AudioTrack {
  Map data;

  bool get isValid =>
      (album != null) && (artist != null) && (path != null) && (title != null);

  String get album => data['album'];

  String get artist => data['artist'];

  String get path => data['path'];

  String get title => data['title'];

  Uint8List get artwork => data['artwork'];

  AudioTrack(
      {String title,
      String artist,
      String album,
      String path,
      Uint8List artwork}) {
    data = {
      'title': title,
      'artist': artist,
      'album': album,
      'path': path,
      'artwork': artwork
    };
  }

  AudioTrack.fromJson(Map data) : data = data;

  String toString() {
    return '$title, $album, $artist';
  }
}

File: src/android/src/main/java/com/twofind/stereo/AudioTrack.java
package com.twofind.stereo;

import android.media.MediaMetadataRetriever;

import java.util.HashMap;

public class AudioTrack {
  private AudioTrack() { }

  public static HashMap<String, Object> toJson(String path) {
    HashMap<String, Object> data = new HashMap<>();

    MediaMetadataRetriever mmr = new MediaMetadataRetriever();
    mmr.setDataSource(path);

    data.put("album", mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUM));
    data.put("artist", mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ARTIST));
    data.put("title", mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_TITLE));
    data.put("path", path);
    data.put("artwork", mmr.getEmbeddedPicture());

    mmr.release();

    return data;
  };
}
